File: src/vigiadev/adapters/stack_detector.py
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

import yaml

from vigiadev.adapters.yaml_parser import dump_config
from vigiadev.domain.errors import ConfigurationError
from vigiadev.domain.models import (
    HealthCheckConfig,
    HealthType,
    PortPolicy,
    ServiceConfig,
    VigiaConfig,
)


class StackDetector:
    """Inspect known project markers without executing project code."""

    COMPOSE_FILES = ("compose.yaml", "docker-compose.yml")

    def __init__(self, root: Path) -> None:
        self.root = root.resolve()
        self._detected: VigiaConfig | None = None
        self._markers: list[str] = []

    def detect(self) -> VigiaConfig:
        services: dict[str, ServiceConfig] = {}
        self._markers = []
        compose_file = self._detect_compose(services)
        self._detect_node(services)
        self._detect_python(services)
        self._detected = VigiaConfig(
            project_name=self._project_name(),
            compose_file=compose_file,
            services=services,
        )
        return self._detected

    def to_yaml(self) -> str:
        config = self._detected or self.detect()
        detected = ", ".join(self._markers) if self._markers else "no known stack markers"
        return dump_config(
            config,
            comments=[
                f"Auto-detected from: {detected}.",
                "Review commands, ports, and healthchecks before sharing this file.",
                "{port} is replaced only for services using port_policy: remap.",
            ],
        )

    def _detect_compose(self, services: dict[str, ServiceConfig]) -> str | None:
        compose_path = next(
            (
                self.root / name
                for name in self.COMPOSE_FILES
                if (self.root / name).is_file()
            ),
            None,
        )
        if compose_path is None:
            return None
        try:
            raw: Any = yaml.safe_load(compose_path.read_text(encoding="utf-8"))
        except (OSError, yaml.YAMLError) as error:
            raise ConfigurationError(f"could not inspect {compose_path}: {error}") from error
        compose_services = raw.get("services") if isinstance(raw, dict) else None
        if not isinstance(compose_services, dict):
            raise ConfigurationError(f"Compose file has no services mapping: {compose_path}")
        for raw_name, definition in compose_services.items():
            name = self._unique_service_name(str(raw_name), services)
            details = definition if isinstance(definition, dict) else {}
            services[name] = ServiceConfig(
                compose_service=str(raw_name),
                ports=self._compose_ports(details.get("ports", [])),
                port_policy=PortPolicy.REUSE,
            )
        self._markers.append(compose_path.name)
        return compose_path.name

    def _detect_node(self, services: dict[str, ServiceConfig]) -> None:
        package_path = self.root / "package.json"
        if not package_path.is_file():
            return
        try:
            package = json.loads(package_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as error:
            raise ConfigurationError(f"could not inspect {package_path}: {error}") from error
        scripts = package.get("scripts", {}) if isinstance(package, dict) else {}
        script = next(
            (
                candidate
                for candidate in ("dev", "start")
                if isinstance(scripts, dict) and isinstance(scripts.get(candidate), str)
            ),
            None,
        )
        if script is None:
            return

        manager = self._node_manager()
        script_body = str(scripts[script]).lower()
        dependencies = {
            **(
                package.get("dependencies", {})
                if isinstance(package.get("dependencies"), dict)
                else {}
            ),
            **(
                package.get("devDependencies", {})
                if isinstance(package.get("devDependencies"), dict)
                else {}
            ),
        }
        dependency_names = {str(name).lower() for name in dependencies}
        port = 5173 if "vite" in script_body or "vite" in dependency_names else 3000
        name = self._unique_service_name("app", services)
        services[name] = ServiceConfig(
            command=self._node_command(manager, script),
            ports=[port],
            port_policy=PortPolicy.REMAP,
            healthcheck=HealthCheckConfig(type=HealthType.TCP, port=port),
        )
        self._markers.append(f"Node.js ({manager}, {script})")

    def _detect_python(self, services: dict[str, ServiceConfig]) -> None:
        manage_py = self.root / "manage.py"
        if manage_py.is_file():
            name = self._unique_service_name("app", services)
            services[name] = ServiceConfig(
                command=["python", "manage.py", "runserver", "0.0.0.0:{port}"],
                ports=[8000],
                port_policy=PortPolicy.REMAP,
                healthcheck=HealthCheckConfig(type=HealthType.TCP, port=8000),
            )
            self._markers.append("Python (Django)")
            return

        dependency_files = [
            path
            for path in (self.root / "pyproject.toml", self.root / "requirements.txt")
            if path.is_file()
        ]
        contents: list[str] = []
        for path in dependency_files:
            try:
                contents.append(path.read_text(encoding="utf-8").lower())
            except OSError as error:
                raise ConfigurationError(f"could not inspect {path}: {error}") from error
        dependency_text = "\n".join(contents)
        if "fastapi" not in dependency_text and "uvicorn" not in dependency_text:
            return
        module = (
            "app"
            if (self.root / "app.py").is_file()
            and not (self.root / "main.py").is_file()
            else "main"
        )
        name = self._unique_service_name("app", services)
        services[name] = ServiceConfig(
            command=[
                "uvicorn",
                f"{module}:app",
                "--host",
                "0.0.0.0",
                "--port",
                "{port}",
            ],
            ports=[8000],
            port_policy=PortPolicy.REMAP,
            healthcheck=HealthCheckConfig(type=HealthType.TCP, port=8000),
        )
        self._markers.append("Python (FastAPI/Uvicorn)")

    def _node_manager(self) -> str:
        for manager, markers in (
            ("pnpm", ("pnpm-lock.yaml",)),
            ("yarn", ("yarn.lock",)),
            ("bun", ("bun.lock", "bun.lockb")),
        ):
            if any((self.root / marker).is_file() for marker in markers):
                return manager
        return "npm"

    @staticmethod
    def _node_command(manager: str, script: str) -> list[str]:
        if manager == "npm":
            return ["npm", "run", script, "--", "--port", "{port}"]
        if manager == "yarn":
            return ["yarn", script, "--port", "{port}"]
        return [manager, "run", script, "--port", "{port}"]

    @staticmethod
    def _compose_ports(raw_ports: Any) -> list[int]:
        if not isinstance(raw_ports, list):
            return []
        ports: list[int] = []
        for entry in raw_ports:
            candidate: Any = entry
            if isinstance(entry, dict):
                candidate = entry.get("published", entry.get("target"))
            elif isinstance(entry, str):
                without_protocol = entry.rsplit("/", 1)[0]
                candidate = (
                    without_protocol.split(":")[-2]
                    if ":" in without_protocol
                    else without_protocol
                )
                candidate = str(candidate).split("-", 1)[0]
            try:
                port = int(candidate)
            except (TypeError, ValueError):
                continue
            if 1 <= port <= 65535 and port not in ports:
                ports.append(port)
        return ports

    def _project_name(self) -> str:
        normalized = re.sub(r"[^A-Za-z0-9_.-]+", "-", self.root.name).strip("._-")
        return normalized or "project"

    @staticmethod
    def _unique_service_name(
        preferred: str, services: dict[str, ServiceConfig]
    ) -> str:
        if preferred not in services:
            return preferred
        suffix = 2
        while f"{preferred}-{suffix}" in services:
            suffix += 1
        return f"{preferred}-{suffix}"

File: internal/adapters/detector/stack_detector.go
package detector

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/Rafael-Albernaz-dev/vigiadev/internal/adapters/docker"
	"github.com/Rafael-Albernaz-dev/vigiadev/internal/domain"
	"gopkg.in/yaml.v3"
)

// StackDetector inspeciona o diretório raiz e sintetiza uma configuração do vigiaDev.
type StackDetector struct {
	RootDir string
}

// NewStackDetector cria um novo detector heurístico apontando para o diretório alvo.
func NewStackDetector(rootDir string) *StackDetector {
	if rootDir == "" {
		rootDir = "."
	}
	return &StackDetector{RootDir: rootDir}
}

// DetectionResult contém a configuração gerada e a lista de marcadores encontrados.
type DetectionResult struct {
	Config  *domain.VigiaConfig
	Markers []string
}

// Detect executa as heurísticas de Docker Compose, Node.js e Python.
func (sd *StackDetector) Detect() (*DetectionResult, error) {
	projectName := filepath.Base(sd.RootDir)
	if projectName == "." || projectName == "/" || projectName == "" {
		projectName = "app"
	}

	cfg := &domain.VigiaConfig{
		Version:     1,
		ProjectName: projectName,
		Services:    make(map[string]domain.ServiceConfig),
		Tasks:       make(map[string]domain.TaskConfig),
	}

	var markers []string

	// 1. Heurística Docker Compose
	composeServices, composeMarker := sd.detectDockerCompose()
	if composeMarker != "" {
		markers = append(markers, composeMarker)
		for name, svc := range composeServices {
			cfg.Services[name] = svc
		}
	}

	// 2. Heurística Node.js
	nodeSvc, nodeMarker := sd.detectNodeJS()
	if nodeMarker != "" {
		markers = append(markers, nodeMarker)
		cfg.Services["frontend"] = nodeSvc
	}

	// 3. Heurística Python
	pythonSvc, pythonMarker := sd.detectPython()
	if pythonMarker != "" {
		markers = append(markers, pythonMarker)
		svcName := "backend"
		if _, exists := cfg.Services["frontend"]; !exists {
			svcName = "app"
		}
		cfg.Services[svcName] = pythonSvc
	}

	tasks, err := sd.DetectTasks()
	if err != nil {
		return nil, err
	}
	cfg.Tasks = tasks

	return &DetectionResult{
		Config:  cfg,
		Markers: markers,
	}, nil
}

// detectDockerCompose procura por compose.yaml ou docker-compose.yml (inclusive com sufixos).
func (sd *StackDetector) detectDockerCompose() (map[string]domain.ServiceConfig, string) {
	composeFile := docker.FindComposeFile(sd.RootDir)
	if composeFile == "" {
		return nil, ""
	}

	foundPath := filepath.Join(sd.RootDir, composeFile)
	data, err := os.ReadFile(foundPath)
	if err != nil {
		return nil, ""
	}

	var composeDoc struct {
		Services map[string]struct {
			Image string        `yaml:"image"`
			Ports []interface{} `yaml:"ports"`
		} `yaml:"services"`
	}

	if err := yaml.Unmarshal(data, &composeDoc); err != nil || len(composeDoc.Services) == 0 {
		return nil, filepath.Base(foundPath)
	}

	services := make(map[string]domain.ServiceConfig)
	for name, svc := range composeDoc.Services {
		var port int
		lowerName := strings.ToLower(name) + " " + strings.ToLower(svc.Image)

		switch {
		case strings.Contains(lowerName, "postgres") || strings.Contains(lowerName, "pgsql"):
			port = 5432
		case strings.Contains(lowerName, "redis"):
			port = 6379
		case strings.Contains(lowerName, "mysql") || strings.Contains(lowerName, "mariadb"):
			port = 3306
		case strings.Contains(lowerName, "mongo"):
			port = 27017
		case strings.Contains(lowerName, "rabbit"):
			port = 5672
		}

		serviceCfg := domain.ServiceConfig{
			ComposeService: name,
			PortPolicy:     domain.PortPolicyReuse,
		}

		if port > 0 {
			serviceCfg.Ports = []int{port}
			serviceCfg.HealthCheck = &domain.HealthCheckConfig{
				Type: domain.HealthCheckTCP,
				Port: port,
			}
		}

		services[name] = serviceCfg
	}

	return services, fmt.Sprintf("Docker Compose (%s)", filepath.Base(foundPath))
}

// detectNodeJS inspeciona package.json e lockfiles.
func (sd *StackDetector) detectNodeJS() (domain.ServiceConfig, string) {
	pkgPath := filepath.Join(sd.RootDir, "package.json")
	data, err := os.ReadFile(pkgPath)
	if err != nil {
		return domain.ServiceConfig{}, ""
	}

	var pkg struct {
		Scripts      map[string]string `json:"scripts"`
		Dependencies map[string]string `json:"dependencies"`
		DevDeps      map[string]string `json:"devDependencies"`
	}
	_ = json.Unmarshal(data, &pkg)

	// Detecta gerenciador de pacotes por lockfile
	pkgManager := sd.packageManager()

	// Identifica script de inicialização
	script := "dev"
	if _, ok := pkg.Scripts["dev"]; !ok {
		if _, ok := pkg.Scripts["start"]; ok {
			script = "start"
		}
	}

	// Infere porta padrão baseada em framework
	port := 3000
	depsStr := fmt.Sprintf("%v %v", pkg.Dependencies, pkg.DevDeps)
	if strings.Contains(depsStr, "vite") {
		port = 5173
	} else if strings.Contains(depsStr, "next") || strings.Contains(depsStr, "nuxt") {
		port = 3000
	}

	cmd := []string{pkgManager, "run", script}
	if pkgManager == "npm" && script == "start" {
		cmd = []string{"npm", "start"}
	}

	svc := domain.ServiceConfig{
		Command:    cmd,
		Ports:      []int{port},
		PortPolicy: domain.PortPolicyRemap,
		HealthCheck: &domain.HealthCheckConfig{
			Type: domain.HealthCheckTCP,
			Port: port,
		},
	}

	return svc, fmt.Sprintf("Node.js (%s run %s)", pkgManager, script)
}

// detectPython inspeciona manage.py (Django) ou FastAPI/uvicorn.
func (sd *StackDetector) detectPython() (domain.ServiceConfig, string) {
	// 1. Django
	if fileExists(filepath.Join(sd.RootDir, "manage.py")) {
		return domain.ServiceConfig{
			Command:    []string{"python", "manage.py", "runserver", "0.0.0.0:{port}"},
			Ports:      []int{8000},
			PortPolicy: domain.PortPolicyRemap,
			HealthCheck: &domain.HealthCheckConfig{
				Type: domain.HealthCheckHTTP,
				URL:  "http://127.0.0.1:{port}/",
			},
		}, "Python (Django manage.py)"
	}

	// 2. FastAPI / Uvicorn
	reqFiles := []string{"requirements.txt", "pyproject.toml"}
	for _, req := range reqFiles {
		p := filepath.Join(sd.RootDir, req)
		if data, err := os.ReadFile(p); err == nil {
			content := strings.ToLower(string(data))
			if strings.Contains(content, "fastapi") || strings.Contains(content, "uvicorn") {
				return domain.ServiceConfig{
					Command:    []string{"uvicorn", "main:app", "--port", "{port}", "--reload"},
					Ports:      []int{8000},
					PortPolicy: domain.PortPolicyRemap,
					HealthCheck: &domain.HealthCheckConfig{
						Type: domain.HealthCheckHTTP,
						URL:  "http://127.0.0.1:{port}/docs",
					},
				}, fmt.Sprintf("Python (FastAPI via %s)", req)
			}
		}
	}

	return domain.ServiceConfig{}, ""
}

// GenerateYAML formata a configuração gerada em um YAML documentado e canônico.
func GenerateYAML(cfg *domain.VigiaConfig, markers []string) (string, error) {
	var sb strings.Builder

	sb.WriteString("# vigiaDev configuration (version 1)\n")
	if len(markers) > 0 {
		sb.WriteString(fmt.Sprintf("# Auto-detectado a partir de: %s\n", strings.Join(markers, ", ")))
	} else {
		sb.WriteString("# Template canônico do vigiaDev\n")
	}
	sb.WriteString("# {port} é interpolado automaticamente caso ocorra colisão (port_policy: remap)\n\n")

	yamlBytes, err := yaml.Marshal(cfg)
	if err != nil {
		return "", err
	}

	sb.Write(yamlBytes)
	return sb.String(), nil
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

// packageManager uses a fixed precedence when multiple lockfiles are present.
func (sd *StackDetector) packageManager() string {
	for _, candidate := range []struct{ file, manager string }{
		{"pnpm-lock.yaml", "pnpm"}, {"yarn.lock", "yarn"},
		{"bun.lockb", "bun"}, {"bun.lock", "bun"}, {"package-lock.json", "npm"},
	} {
		if fileExists(filepath.Join(sd.RootDir, candidate.file)) {
			return candidate.manager
		}
	}
	return "npm"
}

// DetectTasks reads repository metadata only; it never executes discovered commands.
// Explicit Node scripts win generic name collisions, followed by Go and Python.
// Other suites remain accessible as test:go / test:python on a collision.
// Dependencies cannot be inferred reliably from stack markers and remain explicit.
func (sd *StackDetector) DetectTasks() (map[string]domain.TaskConfig, error) {
	tasks := make(map[string]domain.TaskConfig)
	data, err := os.ReadFile(filepath.Join(sd.RootDir, "package.json"))
	if err == nil {
		var pkg struct {
			Scripts map[string]string `json:"scripts"`
		}
		if err := json.Unmarshal(data, &pkg); err != nil {
			return nil, fmt.Errorf("parse package.json: %w", err)
		}
		for name, script := range pkg.Scripts {
			if strings.TrimSpace(name) != "" && strings.TrimSpace(script) != "" {
				tasks[name] = domain.TaskConfig{Command: []string{sd.packageManager(), "run", name}}
			}
		}
	} else if !os.IsNotExist(err) {
		return nil, err
	}
	addSuite := func(stack string, command []string) {
		name := "test"
		if _, exists := tasks[name]; exists {
			name = "test:" + stack
			for {
				if _, exists := tasks[name]; !exists {
					break
				}
				name += ":" + stack
			}
		}
		tasks[name] = domain.TaskConfig{Command: command}
	}
	if fileExists(filepath.Join(sd.RootDir, "go.mod")) {
		addSuite("go", []string{"go", "test", "./..."})
	}
	if fileExists(filepath.Join(sd.RootDir, "manage.py")) {
		addSuite("python", []string{"python", "manage.py", "test"})
	} else {
		pytest := fileExists(filepath.Join(sd.RootDir, "pytest.ini")) || fileExists(filepath.Join(sd.RootDir, "conftest.py"))
		for _, file := range []string{"requirements.txt", "requirements-dev.txt", "pyproject.toml", "setup.cfg", "tox.ini"} {
			data, err := os.ReadFile(filepath.Join(sd.RootDir, file))
			if err != nil && !os.IsNotExist(err) {
				return nil, err
			}
			if strings.Contains(strings.ToLower(string(data)), "pytest") {
				pytest = true
			}
		}
		if pytest {
			addSuite("python", []string{"python", "-m", "pytest"})
		}
	}
	return tasks, nil
}
